### ISSUE TITLE
Surround audio (5.1) export issue

### ISSUE THREAD

**Issue by PurpleAlien** [OpenShot/libopenshot#492]
I've come across a problem where I can not export anything properly with 5.1 surround. The source material are .MTS files (also converted them to mp4 to test, same results). When I export the video, the audio ends before the video clip; feels like it's truncated. I've tried this on the AppImage as well as on Gentoo compiled from source (git).

I see the following in the console when exporting:

[aac @ 0x558a2bc5ed00] Using a PCE to encode channel layout "5.1(side)"

Maybe that has something to do with it?

The source material seems to be identified correctly; from the .osp file:

"reader": {
 "video_stream_index": 0,
 "channel_layout": 1551,
 "video_bit_rate": 3287356,
 "width": 1920,
 "height": 1080,
 "interlaced_frame": false,
 "file_size": "102039552",
 "audio_timebase": {
  "den": 90000,
  "num": 1
 },
 "type": "FFmpegReader",
 "video_length": "1860",
 "pixel_format": 0,
 "sample_rate": 48000,
 "duration": 31.040000915527344,
 "video_timebase": {
  "den": 90000,
  "num": 1
 },
 "channels": 6,
 "metadata": {},
 "path": "./00212.MTS",
 "acodec": "ac3",
 "fps": {
  "den": 1001,
  "num": 60000
 },

**Comment by ferdnyc** [comment]
OK, I have a patch for libopenshot that detects when we're exporting AAC and just silently swaps out `AV_CH_LAYOUT_5POINT1` for `AV_CH_LAYOUT_5POINT1_BACK`. The approach may need to be broadened a little, or we may end up switching to `5POINT1_BACK` as the default across the board, but for now I'm going with a narrowly-targeted fix so we can see whether I'm even on the right track.

I've confirmed that doing that at least eliminates the 'Using a PCE to encode channel layout "5.1(side)"' message. MediaInfo doesn't detect the resulting file any differently, but that's not really that surprising — it's still 5.1 audio either way.

Out of curiosity, how long was the export you were doing when this problem came up, and how long did it go before the audio to cut out?

I've got a 10-minute surround export running now, after it's done I'll repeat it with 2.5.1 release and compare the output. Whether or not I see any difference, as long as it doesn't make the problem any _worse_ I can submit a PR to the repo and get builds generated so I can provide an AppImage that includes this change.

**Comment by ferdnyc** [comment]
Hrm. In both of my exports (2.5.1 release and the current development code with my surround layout change), the audio cuts off at exactly `3 min 32 s 502 ms` into a `10 min 37 s 534 ms` video export. The layout change had no effect on that.

So, there's definitely still a bug here, which I believe I've reproduced, but the cause is _not_ the layout or the use of PCE in the AAC audio stream. At this point I don't know what's causing it, I'll have to run another export with full logging enabled, see if that helps turn up any clues.

**Comment by PurpleAlien** [comment]
Thank you for having a look and being able to confirm! Let me know if you need me to test anything or if I can help in any other way.

**Comment by ferdnyc** [comment]
Update so far: This is the result of my exporting the exact same 10-minute timeline (containing a single 5.1-channel source video)...

1. To MP4 AVC+AAC with stereo audio (top pair of waveforms)
1. To MP4 AVC+AAC in 5.1-channel surround (remaining waveforms)

![image](https://user-images.githubusercontent.com/538020/78176142-0932e400-742a-11ea-9d91-f3ff41a62309.png)

As you can see, the issue is not that the audio is _truncated_, it's that it runs out far too early.

It doesn't appear that the audio is "sped up" per se. From a quick listen, it didn't seem as though pitch was drastically elevated.

Rather, it appears that samples are being dropped **constantly**. In fact, since the output audio is less than half its original length, it would seem that more end up getting dropped than actually make it into the output stream.

cc: @jonoomph @eisneinechse

**Comment by ferdnyc** [comment]
Oh, and if you were wondering whether the exported audio might perhaps be **EXACTLY** one-third the length of the original: That'd be an affirmative.

![image](https://user-images.githubusercontent.com/538020/78177245-d38efa80-742b-11ea-93ad-a5aa69376da1.png)


Clearly, the 5.1-channel processing is mistakenly doing some math somewhere that's based on an assumption of 2-channel output, and is therefore losing 2/3 of the data.

**Comment by ferdnyc** [comment]
(Presumably I could do a 7.1-channel export and come up with a quarter-length version, as well.)

**Comment by SuslikV** [comment]
Side notes (really, the "side").
Some info about surrounding sound (you are not the only one who make mistakes in this):
https://github.com/obsproject/obs-studio/pull/1145#issuecomment-357448369 and few posts below.

**Comment by ferdnyc** [comment]
> I've confirmed that doing that at least eliminates the 'Using a PCE to encode channel layout "5.1(side)"' message. MediaInfo doesn't detect the resulting file any differently, but that's not really that surprising — it's still 5.1 audio either way.

That part actually wasn't true, switching from `LAYOUT_5POINT1` to  `AV_CH_LAYOUT_5POINT1_BACK` actually caused an LFE channel to be detected that wasn't there with the PCE-driven "5.1 (side)" output. MediaInfo detects OpenShot 2.5.1 encodes using 5.1-channel surround like this:
```
Channel(s)                               : 6
Channel(s)                               : 6 channels
Channel positions                        : Front: L C R, Side: C, Back: L R
Channel positions                        : 3/1/2
Channel layout                           : L R C Cb Lb Rb
```

Now, whether in practice that actually means anything, or if it's just a MediaInfo quirk, I couldn't say. But it seems like that switch to `AV_CH_LAYOUT_5POINT1_BACK` might make sense to do even though it's not the source of the incomplete audio output bug.

**Comment by ferdnyc** [comment]
(Interestingly, with the `_BACK` layout, MediaInfo instead detects this — note the **absence** of any rear surround channels — so I'm totally lost now):
```
Channel(s)                               : 6
Channel(s)                               : 6 channels
Channel positions                        : Front: L C R, Side: L R, LFE
Channel positions                        : 3/2/0.1
Channel layout                           : C L R Ls Rs LFE
```

**Comment by PurpleAlien** [comment]
Not sure if this means anything, but in FFmpegWriter.cpp there are two locations in the function write_audio_packets() that set a parameter to AV_SAMPLE_FMT_S16 instead of output_sample_fmt:

int error_code = avcodec_fill_audio_frame(audio_frame, channels_in_frame, AV_SAMPLE_FMT_S16, (uint8_t *) all_queued_samples, all_queued_samples_size, 0);

and

remaining_frame_samples = nb_samples * av_get_bytes_per_sample(AV_SAMPLE_FMT_S16);

Shouldn't those be based on the output_sample_fmt instead of set to AV_SAMPLE_FMT_S16?

I'm not at all familiar with this code base, so I apologize if I'm way off base...

**Comment by ferdnyc** [comment]
@PurpleAlien 

I took a look at that when you brought it up — the sample formats seem to be correct, AFAICT. The places where `AV_SAMPLE_FMT_S16` is used are where it's referencing the _input_ samples, which for that function are (the comments imply) always 16-bit signed. The sample format is separate from the channel count, anyway, so it wouldn't necessarily be different for 5.1-channel vs. 2-channel. Either layout can be using `AV_SAMPLE_FMT_S16` or some other format, for the sample data on **each** channel.

I had my suspicions (and you confirmed) that the sample counts were simply being computed wrong, but it's not because of the sample format (as you also confirmed).

**Comment by ferdnyc** [comment]
One item of particular interest to me is this one, even though it's in a completely different part of the code:
https://github.com/OpenShot/libopenshot/blob/166f69389b6d8a601b51e8d559c418f56a144bba/src/FFmpegWriter.cpp#L913-L919

Note first the comment that even says, "for some reason, it requires me to multiply channels X 2".

But then also note that:
1. Nothing's actually multiplied by 2, though the channel count is multiplied by `av_get_bytes_per_sample(AV_SAMPLE_FMT_S16)` instead
1. It's only done for libavformat >= 54. (Which is ffmpeg 2, so every version we actually support now — still, it's odd).

I don't know if that comment is outdated or what, but it does make me suspect the math in terms of audio data indexing in general. It could be that using the channel count in the code above is actually _incorrect_, so multiplying by `av_get_bytes_per_sample(AV_SAMPLE_FMT_S16)` happens to cancel out for 2-channel, but it ends up dividing by 3 for 5.1-channel.

**Comment by ferdnyc** [comment]
I also notice that ffmpeg's own [`swr_next_pts()`](https://ffmpeg.org/doxygen/4.0/group__lswr.html#gaad9811626c6c01a241f6bf46aff27333) function to deal with those conversions is [_quite_ complex](https://ffmpeg.org/doxygen/4.0/swresample_8c_source.html#l00909), making me think that the ad-hoc version we do just past the section I quoted above may be somewhat lacking. We should also probably just be using their function.

But their code never involves the channel count in any pts math, making me even more suspicious of the quoted section. Take, for example, [`filter_frame`](https://ffmpeg.org/doxygen/4.0/af__aresample_8c_source.html#l00182):
```c
    if(insamplesref->pts != AV_NOPTS_VALUE) {
         int64_t inpts = av_rescale(insamplesref->pts,
             inlink->time_base.num * (int64_t)outlink->sample_rate * inlink->sample_rate,
             inlink->time_base.den);
         int64_t outpts= swr_next_pts(aresample->swr, inpts);
         aresample->next_pts =
         outsamplesref->pts  = ROUNDED_DIV(outpts, inlink->sample_rate);
     } else {
         outsamplesref->pts  = AV_NOPTS_VALUE;
     }
     n_out = swr_convert(aresample->swr, outsamplesref->extended_data, n_out,
                                  (void *)insamplesref->extended_data, n_in);
```

**Comment by ferdnyc** [comment]
`nb_samples` is initially set by the return value of `SWR_CONVERT()`, which is either `swr_convert` or the very deprecated `avresample_convert` depending on the `USE_SWR` define. For some reason we only define that for `libavformat >= 57` which is FFmpeg 3.0, though I don't understand that because libswrescale has existed since at least FFmpeg 2.0, so it seems like we should just always use it. Regardless, let's assume we're dealing with `libswrescale` and therefore the real `swr_convert()`.

What I'm hung up on, right now, is why in hell all this work is done in FFmpegReader and FFmpegWriter to convert the audio data into `AV_SAMPLE_FMT_S16`, when it's going to be stored in the `Frame` as a `juce::AudioSampleBuffer` which is floating-point. (It's a deprecated alias for `juce::AudioBuffer<float>`, to be precise.

FFmpeg _has_ an `AV_SAMPLE_FMT_FLT`, so I am trying very hard to understand why in flippity the data isn't just converted to **_THAT_** instead!?!

...It wouldn't change _this_ issue, or at least I highly doubt it would. But it might make it more clear where the problem _is_ that's causing this issue. The endless conversions and re-conversions of the data are part of what make these functions so impossible to follow. (The other part is that the functions are incredibly, incredibly long spaghetti code built out of cut-and-paste repetitions of the same operations — except _sliiiightly_ different, because the code has morphed over time — and there's absolutely no encapsulation whatsoever.)

Can't do much for performance, either.

**Comment by ferdnyc** [comment]
Whoops, disregard that previous message, it was two comments mixed together. What I meant to say was:

What I'm hung up on, right now, is why in hell all this work is done in FFmpegReader and FFmpegWriter to convert the audio data into `AV_SAMPLE_FMT_S16`, when it's going to be stored in the `Frame` as a `juce::AudioSampleBuffer` which is floating-point. (It's a deprecated alias for `juce::AudioBuffer<float>`, to be precise.

FFmpeg _has_ an `AV_SAMPLE_FMT_FLT`, so I am trying very hard to understand why in flippity the data isn't just converted to **_THAT_** instead!?!

...It wouldn't change _this_ issue, or at least I highly doubt it would. But it might make it more clear where the problem _is_ that's causing this issue. The endless conversions and re-conversions of the data are part of what make these functions so impossible to follow. (The other part is that the functions are incredibly, incredibly long spaghetti code built out of cut-and-paste repetitions of the same operations — except _sliiiightly_ different, because the code has morphed over time — and there's absolutely no encapsulation whatsoever.)

Can't do much for performance, either.

**Comment by PurpleAlien** [comment]
Yes, I've been staring at the code for a few days now and can't understand why it needs to be this complicated. I personally think that a lot of the stuff in FFmpegReader and FFmpegWriter is not needed and the whole thing should be much simpler.

**Comment by SuslikV** [comment]
Programmer is grows together with the code.
Any developers looking back at their own code (after the years were passed) with a wide opened eyes... Pride for algorithm complexity and shame for the obvious bugs.

Big chance that if you'll be able to fix the problem the changes will be accepted. The optimizations should come in separate commit I think.

**Comment by PurpleAlien** [comment]
Would working around the issue by doing something like:

`nb_samples = nb_samples * (info.channels/2);`

be at all acceptable at this time? This fixes things for 5.1 export, and 2 channels still works naturally. I did not test 7.1 or surround (3 channels).
